### PR TITLE
planner/cascades: add transformation rule PushLimitDownOuterJoin

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve? 
This PR adds Transformation rule PushLimitDownOuterJoinwhich is a part of limit in cascades planner(#13709).

### What is changed and how it works?
add a transformation rule PushLimitDownOuterJoin.This rule tries to pushes the Limit through outer Join.

### Check List

Tests 

 - Unit test

Code changes

 - Has exported function/method change

### CODE REVIEW

File: planner/cascades/transformation_rules.go [hunk 1]
```diff
@@ -1415,3 +1416,65 @@ func (r *MergeAdjacentLimit) OnTransform(old *memo.ExprIter) (newExprs []*memo.G
 	newLimitExpr.SetChildren(childGroups...)
 	return []*memo.GroupExpr{newLimitExpr}, true, false, nil
 }
+
+// PushSelDownJoin pushes Limit through Join.
```
Comment (lzmhhh123) [inline planner/cascades/transformation_rules.go]:
```suggestion
// PushLimitDownOuterJoin pushes Limit through Join.
```

File: planner/cascades/transformation_rules.go [hunk 2]
```diff
@@ -1415,3 +1416,65 @@ func (r *MergeAdjacentLimit) OnTransform(old *memo.ExprIter) (newExprs []*memo.G
...
+	join := expr.Children[0].GetExpr().ExprNode.(*plannercore.LogicalJoin)
+	return join.JoinType.IsOuterJoin()
+}
+
+// OnTransform implements Transformation interface.
+// This rule tries to pushes the Limit through outer Join.
+func (r *PushLimitDownOuterJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	limit := old.GetExpr().ExprNode.(*plannercore.LogicalLimit)
+	joinExpr := old.Children[0].GetExpr()
+	join := joinExpr.ExprNode.(*plannercore.LogicalJoin)
```
Comment (lzmhhh123) [inline planner/cascades/transformation_rules.go]: We can merge these two lines by
```
join := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalJoin)
```

File: planner/cascades/transformation_rules.go [hunk 3]
```diff
@@ -1444,3 +1445,64 @@ func (r *TransformLimitToTableDual) OnTransform(old *memo.ExprIter) (newExprs []
...
+	default:
+		return nil, false, false, nil
+	}
+
+	newJoinExpr := memo.NewGroupExpr(join)
+	newJoinExpr.SetChildren(leftGroup, rightGroup)
+	newLimitExpr := memo.NewGroupExpr(limit)
+	newLimitExpr.SetChildren(memo.NewGroupWithSchema(newJoinExpr, joinSchema))
+	newLimitExpr.AddAppliedRule(r)
+	return []*memo.GroupExpr{newLimitExpr}, true, false, nil
```
Comment (zz-jason) [inline planner/cascades/transformation_rules.go]: maybe we can not erase the old plan. take this case as an example:

```sql
select * from s left join t on s.a = t.a limit 10;
```

if the row count of table `s` is smaller than 10, then the new plan `Table Scan(s) -> Limit(10)` is worse than `Table Scan(s)`.

Comment (zz-jason) [reply planner/cascades/transformation_rules.go]: But considered that the `Limit` operator doesn't consume much resource during the execution phase and its execution time is really fast. I think it's OK to greedily erase the old plan:
* it's not that worse in the above case
* it's really worse when the row count of table `s` is so huge and there is not a `Limit(10)` after `Table Scan(s)`

Comment (SeaRise) [reply planner/cascades/transformation_rules.go]: if `limit` be pushed down , `limit->gather->tableScan` will be generate, but not `tableScan->limit`.

Comment (SeaRise) [reply planner/cascades/transformation_rules.go]: What you just said can be considered in `pushLimitDownGather`.
compare `limit.count` and `gather.DataSource.tableStats.RowCount`

Comment (zz-jason) [reply planner/cascades/transformation_rules.go]: > if `limit` be pushed down , `limit->gather->tableScan` will be generate, but not `tableScan->limit`.

Correct. Here `Table Scan(s)` is only a logical speaking. In this rule, I think it should be `Any(on table s) -> Limit(10) -> OuterJoin(s, t) -> Limit(10)`

Comment (zz-jason) [reply planner/cascades/transformation_rules.go]: > What you just said can be considered in `pushLimitDownGather`.
> compare `limit.count` and `gather.DataSource.tableStats.RowCount`

Yes, but without this rule to push the limit down through the join, there can not be a change that matches the pattern `xxGather -> Limit`. these rules are correlated.

Comment (SeaRise) [reply planner/cascades/transformation_rules.go]: Maybe there can add a new attribute `maxRows` in plan like `LogicalPlan.maxRowsPerPartition`.
`gather.maxRows = gather.DataSource.tableStats.RowCount`;
`limit.maxRows = min(child.maxRows, limit.count)`;
`union.maxRows = sum(child.maxRows )`;

`stats.RowCount` is estimated value. `plan.maxRows` is the maximum value

Comment (zz-jason) [reply planner/cascades/transformation_rules.go]: actually we have this mechanism in the execution engine: if the parent only required `maxRows`, the child is better to only return `maxRows`.

**Pull request merged**